pkg/controller/hyperconverged/hyperconverged_controller.go

```diff
@@ -107,261 +109,91 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
-func newKubevirtTemplateValidatorForCR(cr *hcov1alpha1.HyperConverged) *sspv1.KubevirtTemplateValidator {
-	labels := map[string]string{
-		"app": cr.Name,
-	}
-	return &sspv1.KubevirtTemplateValidator{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:   "template-validator-" + cr.Name,
-			Labels: labels,
-		},
+func (r *ReconcileHyperConverged) reconcileUpdate(logger logr.Logger, cr *hcov1alpha1.HyperConverged, request reconcile.Request) (reconcile.Result, error) {
+	c := make(chan ResultFromCRCreate)
+	var wg sync.WaitGroup
+
+	resources := r.getAllResources(cr, request)
+	for _, desiredRuntimeObj := range resources {
+		wg.Add(1)
+		go func(desiredRuntimeObj runtime.Object) {
+			defer wg.Done()
+			desiredMetaObj := desiredRuntimeObj.(metav1.Object)
+
+			// use reflection to create default instance of desiredRuntimeObj type
+			typ := reflect.ValueOf(desiredRuntimeObj).Elem().Type()
+			currentRuntimeObj := reflect.New(typ).Interface().(runtime.Object)
+
+			key := client.ObjectKey{
+				Namespace: desiredMetaObj.GetNamespace(),
+				Name:      desiredMetaObj.GetName(),
+			}
+			err := r.client.Get(context.TODO(), key, currentRuntimeObj)
+
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					c <- ResultFromCRCreate{result: reconcile.Result{}, err: err}
+				}
+
+				if err = controllerutil.SetControllerReference(cr, desiredMetaObj, r.scheme); err != nil {
+					c <- ResultFromCRCreate{result: reconcile.Result{}, err: err}
+				}
+
+				// TODO: common-templates and cdi fails the Get check above and appears to still be missing.
+				// But in reality, when you try to Create it, the client reports back that
+				// the resource already exists. Need to investigate why.
+				// Before the refactor, the code didn't check if a resource already exists. It just
+				// tried to create it, and will skip if the Create indicated that it already exists.
+				if err = r.client.Create(context.TODO(), desiredRuntimeObj); err != nil {
+					if err != nil && errors.IsAlreadyExists(err) {
+						logger.Info("Skip reconcile: tried create but resource already exists", "key", key)
+						c <- ResultFromCRCreate{result: reconcile.Result{}, err: nil}
+					} else if err != nil {
+						c <- ResultFromCRCreate{result: reconcile.Result{}, err: err}
+					}
+				} else {
+					logger.Info("Resource created",
+						"namespace", desiredMetaObj.GetNamespace(),
+						"name", desiredMetaObj.GetName(),
+						"type", fmt.Sprintf("%T", desiredMetaObj))
+				}
+			} else {
+				logger.Info("Skip reconcile: resource already exists", "key", key)
+				c <- ResultFromCRCreate{result: reconcile.Result{}, err: nil}
+			}
+		}(desiredRuntimeObj)
+	}
+
+	go func() {
+		wg.Wait()
+		close(c)
+	}()
+
+	var result reconcile.Result
+	var crErr error
+	var hasError bool
+	// Loop through the results, until an error is found. Once an error is found, don't change result or crErr, but
+	// drain the results channel. If no errors are found, return the last result.
+	for r := range c {
```

Comment:
Will `c` always have all the results from the channel when this is executed?  I'm not sure it will because `wg.Wait()` is run in a thread.

Reply:
Yes, c will always have all the results. wg.Wait() doesn't run until the WaitGroup is drained. That occurs when the last deferred wg.Done() is called when the last goroutine in the loop above completes.

To be honest, I think making this parallel may be overkill at this point. I don't know if we gain that much or at all by making it parallel. It was difficult to assess the difference in speed when I observed the CRs being created before and after this refactor. This may be premature optimization. 

Would you be ok if I rip the parallelism out and but leave the rest in place?

Reply:
I'm not as concerned with speed as I am with resiliency.  As long as we don't continuously fail on the first error we see, it's a big upgrade.  Also, 👍  to more tests.

Reply:
Makes sense to drop the parallelism.  We can come back to it later.

Reply:
Ok removed.